File: 2020/14.py
import re
import random
import copy

def getMask(s):
    mask = []
    for char in s[7:]:
        mask.append(char)
    return mask

def getParams(s):
    params = re.split("\D+", s)
    return int(params[1]), int(params[2])

def decToBin(num):
    binary = []
    binNum = bin(int(num))
    for char in binNum[2:]:
        binary.append(char)
    while len(binary) < 36:
        binary.insert(0, "0")
    return binary

def applyMask(number, mask):
    for i in range(len(mask)):
        if mask[i] == "0":
            number[i] = "0"
        elif mask[i] == "1":
            number[i] = "1"
    return number

def binToDec(num):
    dec = ""
    for char in num:
        dec = dec + char
    return int(dec, 2)

mem = {}
with open("14.txt") as file:
    for line in file:
        if line[0:4] == "mask":
            mask = getMask(line.strip())
        else:
            index, value = getParams(line.strip())
            binNum = decToBin(value)
            newNum = applyMask(binNum, mask)
            decNum = binToDec(newNum)
            mem[index] = decNum

count = 0
for k, v in mem.items():
    count += v
print(count)

def applyMaskv2(number, mask):
    for i in range(len(mask)):
        if mask[i] == "X":
            number[i] = "X"
        elif mask[i] == "1":
            number[i] = "1"
    return number

comb = {0:1}
for i in range(1, 37):
    comb[i] = comb[i-1] * 2

def genIndexes(baseIndex):
    xcount = 0
    for index in baseIndex:
        if index == "X":
            xcount += 1
    maxCombs = comb[xcount]
    indexes = []
    while len(indexes) < maxCombs:
        testIndex = copy.deepcopy(baseIndex)
        for i in range(len(baseIndex)):
            if testIndex[i] == "X":
                testIndex[i] = str(random.randrange(0,2))
        newIndex = binToDec(testIndex)
        if not newIndex in indexes:
            indexes.append(newIndex)
    return indexes

mem = {}
with open("14.txt") as file:
    for line in file:
        if line[0:4] == "mask":
            mask = getMask(line.strip())
        else:
            index, value = getParams(line.strip())
            binIndex = decToBin(index)
            newIndex = applyMaskv2(binIndex, mask)
            indexes = genIndexes(newIndex)
            for number in indexes:
                mem[number] = value

count = 0
for k, v in mem.items():
    count += v
print(count)

File: 2020/Lua/14.lua
local function toBits(num,bits)
    local t = {}
    for b = bits, 1, -1 do
        t[b] = math.fmod(num, 2)
        num = math.floor((num - t[b]) / 2)
    end
    return t
end

local combs = {[0] = 1}
for i = 1, 37 do combs[i] = combs[i - 1] * 2 end

local memory, memory2, mask = {}, {}, nil
for line in io.lines("14.txt") do
  if line:find("mask") then
    mask = line:match("mask = (.+)")
  else
    local dir, n = line:match("mem%[(%d+)%] = (%d+)")
    local bin = toBits(tonumber(n), 36)
    local binDir = toBits(tonumber(dir), 36)
    for i = 1, #mask do
      if mask:sub(i, i) ~= "X" then bin[i] = mask:sub(i,i) end
      if mask:sub(i,i) == "X" or mask:sub(i,i) == "1" then binDir[i] = mask:sub(i,i) end
    end
    memory[tonumber(dir)] = tonumber(table.concat(bin), 2)
    local xCount = 0
    for _, v in pairs(binDir) do if v == "X" then xCount = xCount + 1 end end
    local maxCombs, cmbs, vals = combs[xCount], 0, {}
    while cmbs < maxCombs do
      local s = ""
      for _ = 1, xCount do
        s = s .. math.random(0,1) -- lol
      end
      vals[s] = true
      cmbs = 0
      for _ in pairs(vals) do
        cmbs = cmbs + 1
      end
    end
    for b in pairs(vals) do
      local test, i = table.concat(binDir), 1
      local t = {}
      for c in test:gmatch(".") do table.insert(t, c) end
      for k in pairs(t) do if t[k] == "X" then t[k] = b:sub(i,i) i = i + 1 end end
      memory2[tonumber(table.concat(t), 2)] = tonumber(n)
    end
  end
end

local count = 0
for _, v in pairs(memory) do count = count + v end
print(count)

count = 0
for _, v in pairs(memory2) do count = count + v end
print(count)
